File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
n = int(input())
arr = list(map(int, input().split()))
plus, minus, mult, div = map(int, input().split())

Max = -1e9
Min = 1e9

def func(k, res, plus, minus, mult, div):
        global Max, Min
        if k == n:
            Max = max(res, Max)
            Min = min(res, Min)
            return

        if plus:
            func(k + 1, res + arr[k], plus - 1, minus, mult, div)
        if minus:
            func(k + 1, res - arr[k], plus, minus - 1, mult, div)
        if mult:
            func(k + 1, res * arr[k], plus, minus, mult - 1, div)
        if div:
            func(k + 1, int(res / arr[k]), plus, minus, mult, div - 1)

func(1, arr[0], plus, minus, mult, div)
print(Max)
print(Min)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int plus;
    static int minus;
    static int multi;
    static int divide, n;
    static int[] arr;
    static int max = Integer.MIN_VALUE;
    static int min = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        arr = new int[n];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        plus = Integer.parseInt(st.nextToken());
        minus = Integer.parseInt(st.nextToken());
        multi = Integer.parseInt(st.nextToken());
        divide = Integer.parseInt(st.nextToken());

        func(1, arr[0]);
        System.out.println(max);
        System.out.println(min);
    }

    static void func(int k, int res) {
        if (k == n) {
            max = Math.max(res, max);
            min = Math.min(res, min);
            return;
        }

        if (plus > 0) {
            plus--;
            func(k + 1, res + arr[k]);
            plus++;
        }
        if (minus > 0) {
            minus--;
            func(k + 1, res - arr[k]);
            minus++;
        }
        if (multi > 0) {
            multi--;
            func(k + 1, res * arr[k]);
            multi++;
        }
        if (divide > 0) {
            divide--;
            func(k + 1, res / arr[k]);
            divide++;
        }
    }
}